Sources/_MatchingEngine/Regex/Parse/LexicalAnalysis.swift

```diff
@@ -12,143 +12,140 @@ API convention:
-      throw ParseError.expectedDigits(
-        str, expecting: digitRange)
-    }
-    guard let i = Num(str, radix: 16) else {
-      // TODO: Or, it might have failed because of overflow,
-      // Can we tell easily?
-      throw ParseError.expectedHexNumber(str)
+  /// Try to eat a number of a particular type and radix off the front.
+  ///
+  /// Returns: `nil` if there's no number, otherwise the number
+  ///
+  /// Throws on overflow
+  ///
+  private mutating func lexNumber<Num: FixedWidthInteger>(
+    _ ty: Num.Type, _ kind: RadixKind
+  ) throws -> Value<Num>? {
+    try recordLoc { src in
+      guard let str = src.tryEatPrefix(kind.characterFilter)?.string else {
+        return nil
+      }
+      guard let i = Num(str, radix: kind.radix) else {
+        throw ParseError.numberOverflow(str)
+      }
+      return i
     }
-    return i
   }
 
-
   /// Try to eat a number off the front.
   ///
   /// Returns: `nil` if there's no number, otherwise the number
   ///
   /// Throws on overflow
   ///
   mutating func lexNumber() throws -> Value<Int>? {
-    try recordLoc {
-      try $0.consumeNumber(
-        \.isNumber, validate: { _ in }, radix: 10)
-    }
+    try lexNumber(Int.self, .decimal)
   }
 
   /// Eat a scalar value from hexadecimal notation off the front
   private mutating func expectUnicodeScalar(
     numDigits: Int
   ) throws -> Value<Unicode.Scalar> {
     try recordLoc { src in
-      let num: UInt32 = try src.consumeHexNumber(
-        numDigits: numDigits)
-      guard let scalar = Unicode.Scalar(num) else {
-        throw ParseError.misc(
-          "Invalid scalar value U+\(num.hexStr)")
-      }
-      return scalar
-    }
-  }
-
-  /// Eat a scalar value from hexadecimal notation off the front
-  private mutating func expectUnicodeScalar(
-    digitRange: ClosedRange<Int>
-  ) throws -> Value<Unicode.Scalar> {
-    try recordLoc { src in
-      let uOpt: UInt32? = try src.consumeHexNumber { s in
-        guard digitRange.contains(s.count) else {
-          throw ParseError.expectedDigits(
-            s, expecting: digitRange)
-        }
-      }
-      guard let u = uOpt else {
-        throw ParseError.misc("Expected scalar value")
-      }
-      guard let scalar = Unicode.Scalar(u) else {
-        throw ParseError.misc(
-          "Invalid scalar value U+\(u.hexStr)")
+      let str = src.eat(upTo: numDigits).string
+      guard str.count == numDigits else {
+        throw ParseError.expectedNumDigits(str, numDigits)
       }
-      return scalar
+      return try Source.validateUnicodeScalar(str, .hex)
     }
   }
 
   /// Eat a scalar off the front, starting from after the
   /// backslash and base character (e.g. `\u` or `\x`).
   ///
-  ///     UniScalar -> 'u{' HexDigit{1...8}
+  ///     UniScalar -> 'u{' HexDigit{1...} '}'
   ///                | 'u'  HexDigit{4}
-  ///                | 'x{' HexDigit{1...8}
+  ///                | 'x{' HexDigit{1...} '}'
```

Comment:
CC @natecook1000 

We will likely also want fixits to convert everything to a single canonical notation